src/StreamJsonRpc/ProxyGeneration.cs

```diff
@@ -0,0 +1,353 @@
+                    ImplementEventAccessor(removeMethod.GetILGenerator(), evtField, DelegateRemoveMethod);
+                    evtBuilder.SetRemoveOnMethod(removeMethod);
+
+                    // void OnEventName(EventArgs args)
+                    var eventArgsType = evt.EventHandlerType.GetTypeInfo().GetDeclaredMethod(nameof(EventHandler.Invoke)).GetParameters()[1].ParameterType;
+                    var raiseEventMethod = proxyTypeBuilder.DefineMethod(
+                        $"On{evt.Name}",
+                        MethodAttributes.HideBySig | MethodAttributes.Private,
+                        null,
+                        new Type[] { eventArgsType });
+                    ImplementRaiseEventMethod(raiseEventMethod.GetILGenerator(), evtField, jsonRpcField);
+
+                    ctorActions.Add(new Action<ILGenerator>(il =>
+                    {
+                        var addLocalRpcMethod = typeof(JsonRpc).GetRuntimeMethod(nameof(JsonRpc.AddLocalRpcMethod), new Type[] { typeof(string), typeof(Delegate) });
+                        var delegateCtor = typeof(Action<>).MakeGenericType(eventArgsType).GetTypeInfo().DeclaredConstructors.Single();
+
+                        // rpc.AddLocalRpcMethod("EventName", new Action<EventArgs>(this.OnEventName));
+                        il.Emit(OpCodes.Ldarg_1); // .ctor's rpc parameter
+                        il.Emit(OpCodes.Ldstr, evt.Name);
+                        il.Emit(OpCodes.Ldarg_0);
+                        il.Emit(OpCodes.Ldftn, raiseEventMethod);
+                        il.Emit(OpCodes.Newobj, delegateCtor);
+                        il.Emit(OpCodes.Callvirt, addLocalRpcMethod);
+                    }));
+                }
+
+                // .ctor(JsonRpc)
+                {
+                    var ctor = proxyTypeBuilder.DefineConstructor(MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.SpecialName | MethodAttributes.RTSpecialName, CallingConventions.Standard, new Type[] { typeof(JsonRpc) });
+                    var il = ctor.GetILGenerator();
+
+                    // : base()
+                    il.Emit(OpCodes.Ldarg_0);
+                    il.Emit(OpCodes.Call, ObjectCtor);
+
+                    // this.rpc = rpc;
+                    il.Emit(OpCodes.Ldarg_0);
+                    il.Emit(OpCodes.Ldarg_1);
+                    il.Emit(OpCodes.Stfld, jsonRpcField);
+
+                    // Emit IL that supports events.
+                    foreach (var action in ctorActions)
+                    {
+                        action(il);
+                    }
+
+                    il.Emit(OpCodes.Ret);
+                }
+
+                // IDisposable.Dispose()
+                if (disposable)
+                {
+                    var disposeMethod = proxyTypeBuilder.DefineMethod(nameof(IDisposable.Dispose), MethodAttributes.Private | MethodAttributes.HideBySig | MethodAttributes.NewSlot | MethodAttributes.Virtual);
```

Comment:
The `Dispose()` implementation could be written in C# instead of using `il.Emit()`, by putting it on an abstract base class that is extended by the generated type. The base class could also provide helper methods that could simplify some of the other generated code.

Reply:
True. I noticed that the Cascade implementation derived from a base class. If I had to do anything particularly complicated that I could move to a base class I would consider that. So far, anything that could move to a base class was simple enough that I didn't mind emitting IL for it. That gives me some flexibility to potentially in the future support folks bringing their own base classes to the table if they want, and we'll derive from that. We wouldn't be able to do that if we had to derive from our own base class.

Reply:
OK